twoliter/src/lock.rs

```diff
@@ -139,15 +249,181 @@ impl Lock {
+
+    async fn pull_image<P>(&self, path: P, image: &LockedImage, digest: &str) -> Result<IndexView>
+    where
+        P: AsRef<Path>,
+    {
+        let digest_uri = image.digest_uri(digest);
+        let path = path.as_ref();
+        let digest_file = path.join("digest");
+        let cached_path = path.join("cached");
+        let oci_archive_path = cached_path.join(digest.replace(':', "-"));
+        if !oci_archive_path.exists() {
+            let oci_archive_str = oci_archive_path.to_str().context(format!(
+                "failed to convert path to str: {}",
+                oci_archive_path.display()
+            ))?;
+            docker!(
+                ["save", digest_uri.as_str(), "-o", oci_archive_str],
+                format!(
+                    "failed to fetch kit and save to disk from {} to {}",
+                    digest_uri, oci_archive_str
+                )
+            );
+        }
+        let mut oci_file = File::open(&oci_archive_path).context(format!(
+            "failed to open oci archive at {}",
+            oci_archive_path.display()
+        ))?;
+        let mut oci_archive = Archive::new(&mut oci_file);
+        oci_archive
+            .unpack(path)
+            .context("failed to unpack oci image")?;
+        let index_bytes = read(path.join("index.json")).await?;
+        let index = serde_json::from_slice(index_bytes.as_slice())
+            .context("failed to deserialize oci image index")?;
+        write(&digest_file, digest).await.context(format!(
+            "failed to record digest to {}",
+            digest_file.display()
+        ))?;
+        Ok(index)
+    }
+
+    async fn extract_kit<P>(
+        &self,
+        build_path: P,
+        path: P,
+        image: &LockedImage,
+        arch: &str,
+    ) -> Result<()>
+    where
+        P: AsRef<Path>,
+    {
+        let tmp_dir = TempDir::new_in(build_path)
+            .context("failed to create a temporary directory for kit extraction")?;
+        let vendor = image.vendor.clone();
+        let name = image.name.clone();
+        let docker_arch = match arch {
+            "x86_64" => Ok("amd64"),
+            "aarch64" => Ok("arm64"),
+            _ => Err(anyhow!("unsupported architecture '{}'", arch)),
+        }?;
+        let target_path = path.as_ref().join(format!("{vendor}/{name}/{arch}"));
+        create_dir_all(&target_path).await?;
+
+        // First get the manifest for the specific requested architecture
+        let manifest = self.get_manifest(image, docker_arch).await?;
+
+        // Now pull the image and extract it to a temporary folder returning
+        // the oci index
+        let digest_file = target_path.join("digest");
+        if digest_file.exists()
+            && read_to_string(&digest_file).await.context(format!(
+                "failed to read cached digest id from {}",
+                digest_file.display()
+            ))? == manifest.digest
+        {
+            return Ok(());
+        }
```

Comment:
You end up with this special "digest" marker file in two places that are loosely coupled. It might be better to refactor it in a more object-oriented spirit:
```
let oci_archive = OciArchive::new(&digest);
# checks for the saved image locally, or else pulls and saves it
oci_archive.pull_image()?;
# checks for the matching digest in path
# otherwise cleans up the path and unpacks the archive
oci_archive.unpack_layers(&target_path)?
```